File: 2022/day01/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseInput(filename) {
  const content = readFileSync(filename, 'utf-8').trim();

  return content.split('\n\n').map(group => {
    return group.split('\n')
      .filter(line => line)
      .reduce((sum, line) => sum + parseInt(line, 10), 0);
  });
}

function part1(elves) {
  return Math.max(...elves);
}

function part2(elves) {
  const sorted = [...elves].sort((a, b) => b - a);
  return sorted[0] + sorted[1] + sorted[2];
}

const inputFile = join(__dirname, '..', 'input.txt');
const elves = parseInput(inputFile);

console.log('Part 1:', part1(elves));
console.log('Part 2:', part2(elves));

File: 2022/day01/rust/solution.rs
use std::fs;
use std::path::Path;

fn parse_input(filename: &Path) -> Vec<u64> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");

    content
        .trim()
        .split("\n\n")
        .map(|group| {
            group
                .lines()
                .filter(|line| !line.is_empty())
                .map(|line| line.parse::<u64>().expect("Failed to parse number"))
                .sum()
        })
        .collect()
}

fn part1(elves: &[u64]) -> u64 {
    *elves.iter().max().expect("No elves found")
}

fn part2(elves: &[u64]) -> u64 {
    let mut sorted = elves.to_vec();
    sorted.sort_unstable_by(|a, b| b.cmp(a));
    sorted.iter().take(3).sum()
}

fn main() {
    let exe_path = std::env::current_exe().expect("Failed to get executable path");
    let exe_dir = exe_path.parent().expect("Failed to get executable directory");
    let input_path = exe_dir.join("../input.txt");

    let elves = parse_input(&input_path);

    println!("Part 1: {}", part1(&elves));
    println!("Part 2: {}", part2(&elves));
}
